Implement Fast Paths for most A32 SIMD instructions

riperiperi opened the pull request:
This PR adds x86 intrinsic paths to most of the A32 SIMD instructions. This should bring performance more in line with A64, and the slow/non SSE paths should not be affected.

Since this adds fast paths for most instructions I'll go over the omissions instead. 

Omissions:
- Memory instructions.
- VABS+VNEG integer, VDUP, integer multiply (and fused variants) *
- Vector by scalar for integer *
- Narrow ops (there are only like 2 right now)
- VRECPE.f64, VRSQRTE.f64 (integer is not implemented at all, 64 has no x86 intrinsic)
- VCVT f32<->f64. *
- VCVT f32/64 to integer when using fpscr for rounding mode.
- VCVT integer to f32/64. *
- Anything with round mode "away from zero"
- Move immediate. *
- Shift instructions.
- Integer compare instructions.

Everything else should be covered.

Starred ommissions are ones that don't really have any complicated reason to avoid implementing them. Due to the extract/insert experience you have to go through for using scalars in A32, implementing the scalar forms probably won't be much faster.

Two new intrinsics have been added to the list to allow easier movement of scalar values, since A32 allows you to access them from indexed locations within a V128 and x86 does not:

- insertps
- movss

No tests have been added in this version, though all existing ones (that don't require fastfp = false) should pass. Some tests were added to a32-wip alongside the development of these intrinsics to ensure they were correct, which are currently in master.

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdHelper32.cs:
```diff
@@ -473,6 +473,446 @@ public static void EmitVectorUnaryNarrowOp32(ArmEmitterContext context, Func1I e
+            Debug.Assert(target.Type == OperandType.V128 && value.Type == OperandType.V128);
+
+            int targetSide = targetV & 1;
+            int shuffleMask = 2 | 0;
```
Not sure why the `| 0`, also other methods (well, the one below) seems to directly use a constant value of 2.

riperiperi replied:
I wanted to be explicit about the indices that went in each doubleword, but yeah it just looks silly.

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdHelper32.cs:
```diff
@@ -473,6 +473,446 @@ public static void EmitVectorUnaryNarrowOp32(ArmEmitterContext context, Func1I e
+
+            Operand res = vectorFunc(d, n, m);
+
+            if (!op.Q) //register insert
```
`// Register insert.`

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdMove32.cs:
```diff
@@ -128,209 +144,411 @@ public static void Vtbl(ArmEmitterContext context)
-                    }
-                    
-                    lookupResult = context.ShiftRightUI(lookupResult, subVecIndexShift); // Get the relevant byte from this vector.
+                    (int vn, int en) = GetQuadwordAndSubindex(op.Vn + i, op.RegisterSize);
```
You can just do `tableTuples[i] = GetQuadwordAndSubindex(op.Vn + i, op.RegisterSize);` directly.

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdMove32.cs:
```diff
@@ -128,209 +144,411 @@ public static void Vtbl(ArmEmitterContext context)
+
+                        if (op.Size < 3)
+                        {
+                            d = context.AddIntrinsic(Intrinsic.X86Pshufb, d, mask);
```
I think its better to just add that on the if above to avoid a useless comparison.

riperiperi replied:
I think there was originally something in the middle, just forgot to combine them when I removed it.

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdMove32.cs:
```diff
@@ -128,209 +144,411 @@ public static void Vtbl(ArmEmitterContext context)
+                    Operand dIns, mIns;
+                    if (index >= pairs)
+                    {
+                        int pind = index - pairs;
```
Maybe `pairIndex`?

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdArithmetic32.cs:
```diff
@@ -15,7 +15,28 @@ static partial class InstEmit32
+                    if ((op.Size & 1) == 0)
+                    {
+                        Operand mask = X86GetScalar(context, -0f);
+                        return context.AddIntrinsic(Intrinsic.X86Andnps, mask, m);
```
This pattern is used on a bunch of places actually, might be better to extract it to a separate function and move it to one of those helper static classes.

riperiperi replied:
I'll move it to `InstEmitSimdHelper.EmitFloatAbs`. There are two related uses in A64, and that class is static imported in A32 as well. I'll hold back on modifying A64 beyond the existing uses of this pattern since it's outside the scope of this PR.

riperiperi commented on ARMeilleure/Instructions/InstEmitSimdCmp32.cs:
```diff
@@ -110,7 +160,14 @@ public static void Vclt_Z(ArmEmitterContext context)
-                EmitCmpOpF32(context, SoftFloat32.FPCompareLTFpscr, SoftFloat64.FPCompareLTFpscr, true);
+                if (Optimizations.FastFP && Optimizations.UseSse2)
+                {
+                    EmitSse2CmpOpF32(context, CmpCondition.LessThanOrEqual, true);
```
Good catch, good thing we have reviews 😌

riperiperi commented on ARMeilleure/Instructions/InstEmitSimdHelper32.cs:
```diff
@@ -473,6 +473,446 @@ public static void EmitVectorUnaryNarrowOp32(ArmEmitterContext context, Func1I e
 
+        // Intrinsic Helpers
+
+        public static Operand EmitSwapDoubleWordToSide(ArmEmitterContext context, Operand input, int originalV, int targetV)
```
It was originally going to be a swap, but the optimisation I wanted to do by keeping the old value wasn't really possible. I'll change the name to match.

riperiperi commented on ARMeilleure/Instructions/InstEmitSimdArithmetic32.cs:
```diff
@@ -477,58 +806,115 @@ public static void Vmls_1(ArmEmitterContext context)
+                            return context.AddIntrinsic(Intrinsic.X86Pshufb, op1, mask);
+                    }
+
+                    throw new InvalidOperationException("Unknown VREV Opcode + Size combo.");
```
This is unreachable, the decoder gets size from range -1 to 2 (just due to how the value works, it's 2 - size), and sets undefined when == -1.

gdkchan replied:
Then maybe add a comment `// Should be unreachable.` and change `Unknown` to `Invalid`.

riperiperi replied:
Oops, I was looking at VDUP which has a similar throw. I need to change the opcode for this one to set undefined as defined in the manual:
`if UInt(op)+UInt(size) >= 3 then UNDEFINED;`
I will also change the comment on this.

riperiperi commented on ARMeilleure/Instructions/InstEmitSimdCmp32.cs:
```diff
@@ -224,8 +281,74 @@ private static void EmitVcmpOrVcmpe(ArmEmitterContext context, bool signalNaNs)
+                        context.BitwiseOr(
+                            context.ShiftLeft(zf, Const(2)),
+                            context.ShiftLeft(nf, Const(3))
+                            )
```
I'm not sure why I'm doing it like this anyways, since the flags are split again afterwards. It's likely a leftover from when the FPSCR flags were set using a managed call. I'll add a path for setting with four boolean operands.

riperiperi commented:
Addressed comments, and fixed a bug with VCVT_R where op.Opc2 could be >3 as we were using all 3 bits instead of the lower two. (special thanks to the new exception)

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdCvt32.cs:
```diff
@@ -250,6 +250,8 @@ private static FPRoundingMode Opc2ToRoundMode(int opc2)
                     roundMode = FPRoundingMode.TowardsMinusInfinity;
                     break;
+                default:
+                    throw new ArgumentOutOfRangeException($"Round mode {opc2} out of supported range.");
```
`ArgumentOutOfRangeException` constructor takes a argument name, not a exception message.

riperiperi replied:
I just realized too, this function should probably just do the (opc2 & 3) rather than the callers since it does advertise that it takes Opc2 raw.

riperiperi replied:
Actually, & 3 is useful for the slow paths. I'll rename the function and argument.

riperiperi commented:
A little more remodelling done.

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdCvt32.cs:
```diff
@@ -251,7 +251,7 @@ private static FPRoundingMode Opc2ToRoundMode(int opc2)
                     break;
                 default:
-                    throw new ArgumentOutOfRangeException($"Round mode {opc2} out of supported range.");
+                    throw new ArgumentOutOfRangeException($"rm");
```
`nameof(rm)`

gdkchan replied:
On a side note Im not a huge fan of the name "rm" since on the manual its usually used as the name of the second operand register.

riperiperi replied:
On the manual it is called `RM` in caps, distinct from `Rm` in case only. The function is named approprately, but the naming convention on variables obviously doesn't allow that.

jduncanator replied:
You can pass a message as the second parameter of `ArgumentOutOfRangeException` that has a more detailed message and include `RM` (in caps) in there for distinction.

gdkchan commented on ARMeilleure/Instructions/InstEmitSimdHelper.cs:
```diff
@@ -1467,6 +1467,14 @@ public static Operand EmitBinarySatQAccumulate(ArmEmitterContext context, Operan
 
+        public static Operand EmitFloatAbs(ArmEmitterContext context, Operand value, bool single, bool vector)
+        {
+            Operand mask = vector ? X86GetAllElements(context, -0f) : X86GetScalar(context, -0f);
```
Just a question, is the mask correct for double values? Because it only pass float here regardless of the value of the single argument.

riperiperi replied:
It's not, I just got excited about not having to write a double ternary for the intrinsic selection since we don't have a scalar andn, and forgot about this one.
